### PR TITLE
Qualified constraints (issue #3502) part 2

### PR DESCRIPTION
**Pull request by robjhen**
This is some further refactoring to address issues discussed in PR #4219

### CODE REVIEW

File: cabal-install/Distribution/Client/CmdFreeze.hs [hunk 1]
```diff
@@ -150,7 +150,7 @@ projectFreezeConstraints plan =
     versionConstraints :: Map PackageName [(UserConstraint, ConstraintSource)]
     versionConstraints =
       Map.mapWithKey
-        (\p v -> [(UserConstraint UserUnqualified p (PackagePropertyVersion v),
+        (\p v -> [(UserConstraint UserToplevel p (PackagePropertyVersion v),
```
Comment (grayjay) [inline cabal-install/Distribution/Client/CmdFreeze.hs]: That's a lot clearer.  Thanks for renaming the constructors.

File: cabal-install/Distribution/Client/Targets.hs [hunk 2]
```diff
@@ -729,8 +729,9 @@ readUserConstraint str =
       Just c  -> Right c
   where
     msgCannotParse =
-         "expected a package name followed by a constraint, which is "
-      ++ "either a version range, 'installed', 'source' or flags"
+         "expected a (possibly qualified) package name followed by a " ++
+         "constraint, which is either a version range, 'installed', " ++
+         "'source', 'test', 'bench', or flags"
```
Comment (grayjay) [inline cabal-install/Distribution/Client/Targets.hs]: Good catch!

File: cabal-install/Distribution/Client/Dependency.hs [hunk 3]
```diff
@@ -361,7 +361,7 @@ dontUpgradeNonUpgradeablePackages params =
   where
     extraConstraints =
       [ LabeledPackageConstraint
-        (PackageConstraint (unqualified pkgname) PackagePropertyInstalled)
+        (PackageConstraint (scopeToplevel pkgname) PackagePropertyInstalled)
```
Comment (grayjay) [inline cabal-install/Distribution/Client/Dependency.hs]: The non-upgradable package constraints can use `ScopeAnyQualifier` now.

Comment (robjhen) [inline cabal-install/Distribution/Client/Dependency.hs]: agreed

**Comment by Ericson2314** [comment]
I saw this from a conflicting rebase of some local changes, and figured I'd mention https://github.com/haskell/cabal/issues/1493#issuecomment-272965698 as it's relevant

**Pull request merged**